File: 2026美赛高频算法+创新算法汇总+源代码_副本/50多种常用算法源代码！（Matlab+Python）/Kendall 相关系数（特殊的相关性分析）Matlab+Python代码/kendall_correlation.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Kendall相关系数分析案例说明：
本案例实现Kendall秩相关系数（τ系数）分析，这是一种非参数统计方法，用于衡量
两个变量之间的单调相关程度。与Spearman相关系数相比，Kendall更适用于样本量
较小或存在较多相同秩次（平局）的数据。

Kendall系数取值范围为[-1, 1]：
- 1表示完全正相关
- -1表示完全负相关
- 0表示无单调相关

示例使用汽车数据集，分析汽车重量与油耗之间的相关性，并与Pearson和Spearman
系数进行对比，展示不同相关系数的特点。
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import kendalltau, pearsonr, spearmanr

# 加载数据集
print("加载汽车数据集...")
mpg = sns.load_dataset('mpg')

# 选择两个变量进行分析
# 连续变量：weight（汽车重量）和mpg（每加仑英里数，油耗的反向指标）
x = mpg['weight']
y = mpg['mpg']
var1_name = '汽车重量'
var2_name = '每加仑英里数(mpg)'

# 数据预处理：移除缺失值
data = pd.DataFrame({var1_name: x, var2_name: y}).dropna()
x = data[var1_name]
y = data[var2_name]

# 手动计算Kendall相关系数（简化版，处理小样本）
def kendall_correlation(x, y):
    """手动计算Kendall相关系数"""
    n = len(x)
    concordant = 0  # 一致对数量
    discordant = 0  # 不一致对数量
    
    # 遍历所有数据对
    for i in range(n):
        for j in range(i+1, n):
            # 计算x和y的差异符号
            dx = x[i] - x[j]
            dy = y[i] - y[j]
            
            if dx * dy > 0:
                # 符号相同，一致对
                concordant += 1
            elif dx * dy < 0:
                # 符号相反，不一致对
                discordant += 1
            # 若dx或dy为0，视为平局，不纳入计算
    
    # Kendall系数公式：τ = (一致对 - 不一致对) / C(n,2)
    total_pairs = n * (n - 1) / 2
    if total_pairs == 0:
        return 0
    return (concordant - discordant) / total_pairs

# 计算三种相关系数
kendall_manual = kendall_correlation(x[:50].values, y[:50].values)  # 手动计算（小样本）
kendall_tau, kendall_p = kendalltau(x, y)  # 使用scipy计算
pearson_r, pearson_p = pearsonr(x, y)
spearman_rho, spearman_p = spearmanr(x, y)

# 输出分析结果
print(f"\n{var1_name}与{var2_name}的相关性分析结果：")
print(f"Kendall相关系数（手动计算，小样本）: {kendall_manual:.4f}")
print(f"Kendall相关系数（τ）: {kendall_tau:.4f}，P值: {kendall_p:.8f}")
print(f"Pearson相关系数（r）: {pearson_r:.4f}，P值: {pearson_p:.8f}")
print(f"Spearman相关系数（ρ）: {spearman_rho:.4f}，P值: {spearman_p:.8f}")

# 结果解释
alpha = 0.05
kendall_significant = "显著" if kendall_p < alpha else "不显著"
print(f"\n结果解释：{var1_name}与{var2_name}存在{kendall_significant}的Kendall单调相关")

# 可视化相关性
plt.figure(figsize=(15, 10))

# 1. 散点图与回归线
plt.subplot(2, 2, 1)
sns.regplot(x=x, y=y, scatter_kws={'alpha':0.6}, line_kws={'color':'red'})
plt.title(f'{var1_name}与{var2_name}的散点图及回归线')
plt.xlabel(var1_name)
plt.ylabel(var2_name)
plt.grid(alpha=0.3)

# 2. 三种相关系数对比
plt.subplot(2, 2, 2)
methods = ['Kendall τ', 'Pearson r', 'Spearman ρ']
values = [kendall_tau, pearson_r, spearman_rho]
plt.bar(methods, values, color=['blue', 'orange', 'green'])
plt.axhline(y=0, color='gray', linestyle='--', alpha=0.3)
plt.ylim(-1.1, 1.1)
plt.title('三种相关系数对比')
plt.grid(axis='y', alpha=0.3)

# 3. 秩次散点图（用于Kendall和Spearman）
plt.subplot(2, 2, 3)
x_rank = x.rank()
y_rank = y.rank()
plt.scatter(x_rank, y_rank, alpha=0.6, color='purple')
sns.regplot(x=x_rank, y=y_rank, scatter=False, line_kws={'color':'darkred'})
plt.title(f'{var1_name}与{var2_name}的秩次散点图')
plt.xlabel(f'{var1_name}的秩次')
plt.ylabel(f'{var2_name}的秩次')
plt.grid(alpha=0.3)

# 4. 相关性热图
plt.subplot(2, 2, 4)
corr_df = pd.DataFrame({var1_name: x, var2_name: y})
corr_matrix = corr_df.corr(method='kendall')
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', vmin=-1, vmax=1, 
            center=0, annot_kws={"size": 12})
plt.title('Kendall相关系数热图')

plt.tight_layout()
plt.show()

# 补充说明不同相关系数的适用场景
print("\n相关系数适用场景说明：")
print("- Kendall τ：适用于小样本、存在较多平局或需要更稳健结果的情况")
print("- Pearson r：适用于线性关系且数据近似正态分布的情况")
print("- Spearman ρ：适用于非线性但单调的关系，对异常值较稳健")
    

File: 2026美赛高频算法+创新算法汇总+源代码_副本/50多种常用算法源代码！（Matlab+Python）/Kendall 相关系数（特殊的相关性分析）Matlab+Python代码/kendall_correlation.m
% Kendall相关系数分析案例说明：
% 本案例实现Kendall秩相关系数（τ系数）分析，这是一种非参数统计方法，用于衡量
% 两个变量之间的单调相关程度。与Spearman相关系数相比，Kendall更适用于样本量
% 较小或存在较多相同秩次（平局）的数据。
%
% Kendall系数取值范围为[-1, 1]：
% - 1表示完全正相关
% - -1表示完全负相关
% - 0表示无单调相关
%
% 示例使用汽车数据集，分析汽车重量与油耗之间的相关性，并与Pearson和Spearman
% 系数进行对比，展示不同相关系数的特点。

% 检查统计工具箱
if ~license('test', 'Statistics_and_Machine_Learning_Toolbox')
    error('需要Statistics and Machine Learning Toolbox才能运行完整分析');
end

% 加载数据集
disp('加载汽车数据集...');
try
    % 尝试加载内置数据集
    load mpg.mat;
    x = mpg.Weight;       % 汽车重量
    y = mpg.MPG;          % 每加仑英里数
    var1_name = '汽车重量';
    var2_name = '每加仑英里数(mpg)';
catch
    % 生成模拟数据（汽车重量与油耗负相关）
    rng(42);  % 设置随机种子
    n = 300;
    x = 2000 + 3000 * rand(n, 1);  % 汽车重量：2000-5000磅
    y = 40 - 0.005 * x + 3 * randn(n, 1);  % 每加仑英里数（与重量负相关）
    var1_name = '汽车重量';
    var2_name = '每加仑英里数(mpg)';
end

% 数据预处理：移除缺失值
valid_idx = ~isnan(x) & ~isnan(y);
x = x(valid_idx);
y = y(valid_idx);

% 手动计算Kendall相关系数（简化版，处理小样本）
function tau = kendall_correlation(x, y)
    n = length(x);
    concordant = 0;  % 一致对数量
    discordant = 0;  % 不一致对数量
    
    % 遍历所有数据对
    for i = 1:n
        for j = i+1:n
            % 计算x和y的差异符号
            dx = x(i) - x(j);
            dy = y(i) - y(j);
            
            if dx * dy > 0
                % 符号相同，一致对
                concordant = concordant + 1;
            elseif dx * dy < 0
                % 符号相反，不一致对
                discordant = discordant + 1;
            end
            % 若dx或dy为0，视为平局，不纳入计算
        end
    end
    
    % Kendall系数公式：τ = (一致对 - 不一致对) / C(n,2)
    total_pairs = n * (n - 1) / 2;
    if total_pairs == 0
        tau = 0;
    else
        tau = (concordant - discordant) / total_pairs;
    end
end

% 计算三种相关系数
kendall_manual = kendall_correlation(x(1:50), y(1:50));  % 手动计算（小样本）
[kendall_tau, kendall_p] = corr(x, y, 'Type', 'Kendall');  % Matlab函数计算
[pearson_r, pearson_p] = corr(x, y, 'Type', 'Pearson');
[spearman_rho, spearman_p] = corr(x, y, 'Type', 'Spearman');

% 输出分析结果
fprintf('\n%s与%s的相关性分析结果：\n', var1_name, var2_name);
fprintf('Kendall相关系数（手动计算，小样本）: %.4f\n', kendall_manual);
fprintf('Kendall相关系数（τ）: %.4f，P值: %.8f\n', kendall_tau, kendall_p);
fprintf('Pearson相关系数（r）: %.4f，P值: %.8f\n', pearson_r, pearson_p);
fprintf('Spearman相关系数（ρ）: %.4f，P值: %.8f\n', spearman_rho, spearman_p);

% 结果解释
alpha = 0.05;
if kendall_p < alpha
    kendall_conclusion = sprintf('显著的Kendall单调相关（P=%.8f < %.2f）', kendall_p, alpha);
else
    kendall_conclusion = sprintf('无显著的Kendall单调相关（P=%.8f ≥ %.2f）', kendall_p, alpha);
end
fprintf('\n结果解释：%s与%s存在%s\n', var1_name, var2_name, kendall_conclusion);

% 可视化相关性
figure('Position', [100 100 1000 800]);

% 1. 散点图与回归线
subplot(2, 2, 1);
scatter(x, y, 'b', 'MarkerFaceColor', 'b', 'AlphaData', 0.6);
hold on;
% 添加回归线
p = polyfit(x, y, 1);
y_fit = polyval(p, x);
plot(x, y_fit, 'r', 'LineWidth', 2);
title(sprintf('%s与%s的散点图及回归线', var1_name, var2_name), 'FontSize', 12);
xlabel(var1_name);
ylabel(var2_name);
grid on;
hold off;

% 2. 三种相关系数对比
subplot(2, 2, 2);
methods = {'Kendall τ', 'Pearson r', 'Spearman ρ'};
values = [kendall_tau, pearson_r, spearman_rho];
bar(values, 'FaceColor', 'flat', 'CData', [0.2 0.4 0.6; 0.8 0.4 0; 0 0.6 0.2]);
set(gca, 'XTickLabel', methods, 'XTick', 1:3, 'YLim', [-1.1 1.1]);
title('三种相关系数对比', 'FontSize', 12);
grid on;
yline(0, 'k--', 'LineWidth', 1);

% 3. 秩次散点图（用于Kendall和Spearman）
subplot(2, 2, 3);
x_rank = tiedrank(x);
y_rank = tiedrank(y);
scatter(x_rank, y_rank, 'm', 'MarkerFaceColor', 'm', 'AlphaData', 0.6);
hold on;
% 添加秩次回归线
p_rank = polyfit(x_rank, y_rank, 1);
y_rank_fit = polyval(p_rank, x_rank);
plot(x_rank, y_rank_fit, 'darkred', 'LineWidth', 2);
title(sprintf('%s与%s的秩次散点图', var1_name, var2_name), 'FontSize', 12);
xlabel(sprintf('%s的秩次', var1_name));
ylabel(sprintf('%s的秩次', var2_name));
grid on;
hold off;

% 4. 相关性热图
subplot(2, 2, 4);
corr_matrix = [1 kendall_tau; kendall_tau 1];
heatmap(corr_matrix, ...
        'RowLabels', {var1_name, var2_name}, ...
        'ColumnLabels', {var1_name, var2_name}, ...
        'ColorbarVisible', 'off');
title('Kendall相关系数热图', 'FontSize', 12);

tight_layout;
    